File: ontrack-ui-graphql/src/main/java/net/nemerosa/ontrack/graphql/schema/AbstractGQLProjectEntity.java
package net.nemerosa.ontrack.graphql.schema;

import graphql.schema.*;
import net.nemerosa.ontrack.graphql.support.GraphqlUtils;
import net.nemerosa.ontrack.model.structure.ProjectEntity;
import net.nemerosa.ontrack.model.structure.ProjectEntityType;
import net.nemerosa.ontrack.model.structure.Signature;

import java.util.*;
import java.util.stream.Collectors;

import static graphql.schema.GraphQLFieldDefinition.newFieldDefinition;

public abstract class AbstractGQLProjectEntity<T extends ProjectEntity> implements GQLType {

    public static final String PROJECT_ENTITY = "ProjectEntity";

    private final Class<T> projectEntityClass;
    private final ProjectEntityType projectEntityType;
    private final List<GQLProjectEntityFieldContributor> projectEntityFieldContributors;
    private final GQLTypeCreation creation;

    public AbstractGQLProjectEntity(
            Class<T> projectEntityClass,
            ProjectEntityType projectEntityType,
            List<GQLProjectEntityFieldContributor> projectEntityFieldContributors,
            GQLTypeCreation creation
    ) {
        this.projectEntityClass = projectEntityClass;
        this.projectEntityType = projectEntityType;
        this.projectEntityFieldContributors = projectEntityFieldContributors;
        this.creation = creation;
    }

    protected GraphQLInterfaceType projectEntityInterface() {
        return GraphQLInterfaceType.newInterface()
                .name(PROJECT_ENTITY)
                // Common fields
                .fields(baseProjectEntityInterfaceFields())
                // TODO Type resolver not set, but it should
                .typeResolver(new TypeResolverProxy())
                // OK
                .build();
    }

    protected List<GraphQLFieldDefinition> projectEntityInterfaceFields() {
        List<GraphQLFieldDefinition> definitions = baseProjectEntityInterfaceFields();
        // For all contributors
        definitions.addAll(
                projectEntityFieldContributors.stream()
                        .map(contributor -> contributor.getFields(projectEntityClass, projectEntityType))
                        .filter(Objects::nonNull)
                        .flatMap(Collection::stream)
                        .collect(Collectors.toList())
        );
        // OK
        return definitions;
    }

    private List<GraphQLFieldDefinition> baseProjectEntityInterfaceFields() {
        return new ArrayList<>(
                Arrays.asList(
                        GraphqlUtils.idField(),
                        GraphqlUtils.nameField(),
                        GraphqlUtils.descriptionField(),
                        newFieldDefinition()
                                .name("creation")
                                .type(creation.getType())
                                .dataFetcher(creationFetcher())
                                .build()
                )
        );
    }

    protected DataFetcher creationFetcher() {
        return GraphqlUtils.fetcher(
                projectEntityClass,
                entity -> getSignature(entity)
                        .map(GQLTypeCreation::getCreationFromSignature)
                        .orElse(null)
        );
    }

    protected abstract Optional<Signature> getSignature(T entity);

}

File: ontrack-ui-graphql/src/main/java/net/nemerosa/ontrack/graphql/schema/AbstractGQLProjectEntity.kt
package net.nemerosa.ontrack.graphql.schema

import graphql.Scalars
import graphql.schema.GraphQLFieldDefinition
import graphql.schema.GraphQLTypeReference
import net.nemerosa.ontrack.extension.api.ExtensionManager
import net.nemerosa.ontrack.extension.api.ProjectEntityUserMenuItemExtension
import net.nemerosa.ontrack.graphql.support.descriptionField
import net.nemerosa.ontrack.graphql.support.idField
import net.nemerosa.ontrack.graphql.support.listType
import net.nemerosa.ontrack.graphql.support.nameField
import net.nemerosa.ontrack.model.structure.ProjectEntity
import net.nemerosa.ontrack.model.structure.ProjectEntityType
import net.nemerosa.ontrack.model.structure.Signature
import net.nemerosa.ontrack.model.support.FreeTextAnnotatorContributor
import net.nemerosa.ontrack.model.support.MessageAnnotationUtils

abstract class AbstractGQLProjectEntity<T : ProjectEntity>(
    private val projectEntityClass: Class<T>,
    private val projectEntityType: ProjectEntityType,
    private val projectEntityFieldContributors: List<GQLProjectEntityFieldContributor>,
    private val creation: GQLTypeCreation,
    private val freeTextAnnotatorContributors: List<FreeTextAnnotatorContributor>,
    private val extensionManager: ExtensionManager,
) : GQLType {

    private val projectEntityUserMenuItemExtensions: Set<ProjectEntityUserMenuItemExtension> by lazy {
        extensionManager.getExtensions(ProjectEntityUserMenuItemExtension::class.java).toSet()
    }

    protected fun projectEntityInterfaceFields(): List<GraphQLFieldDefinition> {
        val definitions = baseProjectEntityInterfaceFields().toMutableList()
        // For all contributors
        definitions.addAll(
            projectEntityFieldContributors
                .mapNotNull { contributor: GQLProjectEntityFieldContributor ->
                    contributor.getFields(
                        projectEntityClass,
                        projectEntityType
                    )
                }
                .flatten()
        )
        // OK
        return definitions
    }

    private fun baseProjectEntityInterfaceFields(): List<GraphQLFieldDefinition> =
        listOf(
            idField(),
            nameField(),
            descriptionField(),
            creationField(),
            annotatedDescriptionField(),
            userMenuActionsField(),
        )

    private fun userMenuActionsField(): GraphQLFieldDefinition =
        GraphQLFieldDefinition.newFieldDefinition()
            .name("userMenuActions")
            .description("List of actions available for this entity")
            .type(listType(GQLTypeUserMenuAction.TYPE))
            .dataFetcher { env ->
                val entity: T = env.getSource()
                projectEntityUserMenuItemExtensions.flatMap { extension ->
                    extension.getItems(entity)
                }
            }
            .build()

    private fun annotatedDescriptionField(): GraphQLFieldDefinition =
        GraphQLFieldDefinition.newFieldDefinition()
            .name("annotatedDescription")
            .type(Scalars.GraphQLString)
            .description("Description with links.")
            .dataFetcher { env ->
                val entity: T = env.getSource()
                val description: String? = entity.description
                if (description.isNullOrBlank()) {
                    ""
                } else {
                    val annotators = freeTextAnnotatorContributors.flatMap { it.getMessageAnnotators(entity) }
                    MessageAnnotationUtils.annotate(description, annotators)
                }
            }
            .build()

    private fun creationField(): GraphQLFieldDefinition =
        GraphQLFieldDefinition.newFieldDefinition()
            .name("creation")
            .type(creation.typeRef)
            .dataFetcher { env ->
                val entity: T = env.getSource()
                val signature = getSignature(entity)
                signature?.let {
                    GQLTypeCreation.getCreationFromSignature(it)
                }
            }
            .build()

    protected abstract fun getSignature(entity: T): Signature?

}